File: src/serving_modes.py
"""
Lantern OS Serving Modes: Fast (default) vs. Deep (opt-in).

FAST MODE (product default):
  - Uses cached KV inference (UniversalTransformerCache in Ollama/Ouro)
  - Anti-repetition decode params enabled
  - Target: sub-2s replies for dream chat
  - Suitable for: interactive use, UX feedback, real-time systems

DEEP MODE (opt-in via OURO_NATIVE=1):
  - Native Σ₀ Q-exit loop (adaptive depth, grounded reasoning)
  - Higher reasoning latency (70-85s acceptable for research)
  - Suitable for: architecture decisions, grant writing, core system design

Detection:
  - OURO_NATIVE=1 in env → use deep mode
  - Otherwise → use fast cached mode (default)

Configuration per provider is immutable once set. Switching modes requires server restart.
"""

import os
from dataclasses import dataclass
from typing import Optional


@dataclass
class ServingMode:
    """Describes inference behavior for a request."""
    name: str
    use_kv_cache: bool
    max_latency_ms: int
    decode_antirepetition: bool
    reasoning_depth: str  # "shallow" or "adaptive"
    description: str


FAST_MODE = ServingMode(
    name="fast",
    use_kv_cache=True,
    max_latency_ms=2000,
    decode_antirepetition=True,
    reasoning_depth="shallow",
    description="Cached KV inference with anti-repetition decode. Product default for interactive use.",
)

DEEP_MODE = ServingMode(
    name="deep",
    use_kv_cache=False,
    max_latency_ms=120000,
    decode_antirepetition=True,
    reasoning_depth="adaptive",
    description="Native Σ₀ Q-exit loop. Opt-in for architecture decisions and grounded reasoning.",
)


def get_serving_mode() -> ServingMode:
    """Determine which serving mode to use."""
    if os.environ.get("OURO_NATIVE", "").lower() in ("1", "true", "yes"):
        return DEEP_MODE
    return FAST_MODE


def serving_mode_for(dilation: Optional[float] = None) -> ServingMode:
    """Pick the serving mode for a given time-dilation — the within→without bridge.

    `OURO_NATIVE=1` still hard-forces DEEP. Otherwise escalate FAST→DEEP only when the
    dilation is high enough that `grounding_policy(D).deep_mode` is set (route by
    difficulty: a productively-uncertain turn earns the slow Σ₀ loop; a routine one
    stays on the sub-2s cached path). `dilation=None` ⇒ the env-based default.
    """
    base = get_serving_mode()
    if base.name == "deep" or dilation is None:
        return base
    try:
        from convergence_io.dilation import grounding_policy
        if grounding_policy(float(dilation)).deep_mode:
            return DEEP_MODE
    except Exception:
        pass
    return base


def get_decode_params(mode: ServingMode) -> dict:
    """Get decode parameters appropriate to the serving mode."""
    if not mode.decode_antirepetition:
        return {}

    # Fast mode: aggressive antirepetition to prevent token loops
    if mode.name == "fast":
        return {
            "top_p": 0.95,
            "frequency_penalty": 0.5,
            "repetition_penalty": 1.1,  # For Ollama
            "repeat_last_n": 64,         # For Ollama
        }

    # Deep mode: moderate antirepetition (adaptive loop may need some repetition for grounding)
    if mode.name == "deep":
        return {
            "top_p": 0.98,
            "frequency_penalty": 0.2,
            "repetition_penalty": 1.05,
            "repeat_last_n": 128,
        }

    return {}


def describe_mode() -> str:
    """Human-readable description of current mode."""
    mode = get_serving_mode()
    return f"[{mode.name.upper()}] {mode.description} (max {mode.max_latency_ms}ms)"


if __name__ == "__main__":
    mode = get_serving_mode()
    print(f"Current serving mode: {mode.name}")
    print(f"Description: {mode.description}")
    print(f"Decode params: {get_decode_params(mode)}")

File: apps/lantern-garage/lib/serving-modes.js
"use strict";
/**
 * Serving modes for the dream-chat Node path — FAST (default) vs DEEP (OURO_NATIVE=1).
 *
 * Node mirror of src/serving_modes.py (PR #723 / issue #729). PR #723 landed the
 * anti-repetition decode params in the Python unified_agent_connector, but the live
 * dream-chat product (apps/lantern-garage) builds its own provider requests in
 * dream-chat.js / stream-chat.js and never picked them up. This module closes that
 * gap so the actual product gets the same token-loop mitigation.
 *
 * Coverage matches the merged reference: Ollama + OpenAI-compatible providers
 * (OpenAI / Groq / Deepseek / xAI). The Python connector intentionally leaves
 * Anthropic (no frequency_penalty) and Gemini unmodified, so this module does too.
 *
 * FAST mode (product default): aggressive anti-repetition decode params
 *   (top_p=0.95, frequency_penalty=0.5 for OpenAI-style; repeat_penalty=1.1,
 *   repeat_last_n=64 for Ollama). Target: sub-2s interactive replies, no ✅✅✅ loops.
 * DEEP mode (opt-in via OURO_NATIVE=1): gentler params to allow grounded reasoning.
 *
 * Keep the param values in sync with src/serving_modes.py.
 */

const FAST_MODE = Object.freeze({
  name: "fast",
  maxLatencyMs: 2000,
  description: "Cached inference with anti-repetition decode. Product default for interactive use.",
});

const DEEP_MODE = Object.freeze({
  name: "deep",
  maxLatencyMs: 120000,
  description: "Native Σ₀ Q-exit loop. Opt-in for architecture decisions and grounded reasoning.",
});

/** Return the active serving mode. DEEP only when OURO_NATIVE is truthy. */
function getServingMode() {
  return /^(1|true|yes)$/i.test(process.env.OURO_NATIVE || "") ? DEEP_MODE : FAST_MODE;
}

/**
 * Provider-neutral anti-repetition decode params for a mode.
 * Mirrors get_decode_params() in src/serving_modes.py.
 */
function getDecodeParams(mode) {
  const m = mode || getServingMode();
  if (m.name === "deep") {
    return { top_p: 0.98, frequency_penalty: 0.3, repetition_penalty: 1.1, repeat_last_n: 256 };
  }
  // fast (default). Strengthened for the local-Ollama degraded path (#1609): small
  // served models spiral into mid-generation multi-word / multi-language repetition,
  // and repeat_penalty=1.1 over only the last 64 tokens was too weak to catch it.
  // Raise the penalty, widen the look-back window to cover longer drift, and tighten
  // top_p. Stays within Ollama's native repeat_penalty/repeat_last_n levers (Ollama
  // does not honor frequency_penalty, so we don't fake it there).
  return { top_p: 0.92, frequency_penalty: 0.6, repetition_penalty: 1.18, repeat_last_n: 256 };
}

/**
 * Apply Ollama-shaped decode params onto an `options` object (mutates + returns it).
 * Ollama uses repeat_penalty / repeat_last_n rather than frequency_penalty.
 */
// Stop sequences that cut a local instruction/chat-tuned model's output the moment
// it tries to start a *new* turn or echo its prompt template — the #1 reliability
// problem with served local models (e.g. ouro:latest answers correctly, then appends
// "### Response:" and rambles into a fresh instruction block). These are turn/template
// markers, not content, so they never appear mid-answer in a clean reply.
const OLLAMA_STOP = Object.freeze([
  "### Response:", "### Instruction:", "### Input:", "### Task:",
  "<|im_end|>", "<|endoftext|>", "<|eot_id|>",
  "\n\nUser:", "\n\nHuman:", "\n\nAssistant:", "\n\nQuestion:",
]);

function applyOllamaDecodeParams(options, mode) {
  const opts = options || {};
  const m = mode || getServingMode();
  const dp = getDecodeParams(m);
  opts.top_p = dp.top_p;
  opts.repeat_penalty = dp.repetition_penalty;
  opts.repeat_last_n = dp.repeat_last_n;
  // Reliability over a small local model (Σ₀ thesis: grounded + clean beats raw size):
  // a steadier temperature so it doesn't wander, a length ceiling so a runaway can't
  // ramble forever, and stop sequences so it ends cleanly after the answer instead of
  // echoing its template. Caller-supplied values win.
  if (opts.temperature == null) opts.temperature = m.name === "deep" ? 0.6 : 0.4;
  if (opts.num_predict == null) opts.num_predict = m.name === "deep" ? 2048 : 768;
  opts.stop = Array.isArray(opts.stop) ? Array.from(new Set([...opts.stop, ...OLLAMA_STOP])) : OLLAMA_STOP.slice();
  return opts;
}

/**
 * Apply OpenAI-compatible decode params (OpenAI / Groq / Deepseek / xAI) onto a
 * request body (mutates + returns it): top_p + frequency_penalty.
 */
function applyOpenAIDecodeParams(body, mode) {
  const b = body || {};
  const dp = getDecodeParams(mode);
  b.top_p = dp.top_p;
  b.frequency_penalty = dp.frequency_penalty;
  return b;
}

module.exports = {
  FAST_MODE,
  DEEP_MODE,
  getServingMode,
  getDecodeParams,
  applyOllamaDecodeParams,
  applyOpenAIDecodeParams,
};
